File: 0560 subarray-sum-equals-k/solution.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> map;
        map[0]++;
        int curr = 0,ans = 0;
        for(int i:nums) {
            curr+=i;
            int need = curr - k;
            if(map.count(need)) ans+=map[need];
            map[curr]++;
        }
        return ans;
    }
};

File: 0560 subarray-sum-equals-k/solution.rs
use std::collections::HashMap;
impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut map = HashMap::new();
        map.insert(0,1);
        let mut curr = 0;
        let mut ans = 0;
        for i in nums {
            curr+=i;
            let need = curr-k;
            if let Some(&count) = map.get(&need) {
                ans += count;
            }
            *map.entry(curr).or_insert(0) += 1;
        }
        ans
    }
}
